### DISCUSSION TITLE
reason why semaphore cannot be initialized to negative?

### DISCUSSION THREAD

**Discussion by camerondavison** [agronholm/anyio#413] [Q&A]
I see
```
        if value < 0:
            raise ValueError("Semaphore initial value must be >= 0")
```
but I think that it is a common threading construct to initialize a semaphore to a negative number to work as a barrier for the next task to complete.

```python
async def task(s, i):
   sleep(i)
   s.release()

# wait for any 3 out of the 5 tasks to finish
s = Semaphore(-3)
async with create_task_group() as tg:
  for i in range(5):
    tg.start_soon(task, s, i)
  await s.acquire()
```

quickly glancing through the code, it seems like it would work with a negative number

**Comment by TayfurYldz** [comment]
The restriction is intentional because AnyIO has to preserve the semaphore contract across its supported backends, and a negative initial value is not portable there.

In particular, the asyncio backend mirrors `asyncio.Semaphore`, which rejects negative initial values, while the Trio backend delegates to `trio.Semaphore`. Allowing a negative value only in AnyIO's own bookkeeping would therefore give different semantics depending on the selected backend.

There is also a subtle issue with using a negative semaphore as “wait for N completions”: a normal semaphore acquisition means “consume one currently available permit”. It does not represent a countdown threshold. Conflating those two meanings makes `value`, `acquire_nowait()`, bounded-semaphore behavior, and backend interoperability harder to define.

For the example here, I would model the condition explicitly instead. Have each worker increment a completion counter under an AnyIO `Condition` (or signal a memory object stream), and have the waiter proceed when the counter reaches 3. A zero-capacity/normal semaphore can also be used by releasing once per completed worker and acquiring three times:

```python
done = Semaphore(0)

async def task(i):
    await sleep(i)
    done.release()

async with create_task_group() as tg:
    for i in range(5):
        tg.start_soon(task, i)

    for _ in range(3):
        await done.acquire()
```

That expresses “wait for any three completions” without relying on negative semaphore semantics, and it works consistently across AnyIO backends.

One small correction to the original sketch: `sleep(i)` also needs to be awaited.
